x/pool-incentives/keeper/grpc_query.go

```diff
@@ -143,54 +143,45 @@ func (q Querier) IncentivizedPools(ctx context.Context, _ *types.QueryIncentiviz
 	// While there are exceptions, typically the number of incentivizedPools
 	// equals to the number of incentivized gauges / number of lockable durations.
 	incentivizedPools := make([]types.IncentivizedPool, 0, len(distrInfo.Records)/len(lockableDurations))
-	incentivizedPoolIDs := make(map[uint64]time.Duration)
 
 	// Loop over the distribution records and fill in the incentivized pools struct.
 	for _, record := range distrInfo.Records {
-		for _, lockableDuration := range lockableDurations {
-			poolId, err := q.Keeper.GetPoolIdFromGaugeId(sdkCtx, record.GaugeId, lockableDuration)
-			if err == nil {
-				incentivizedPool := types.IncentivizedPool{
-					PoolId:           poolId,
-					LockableDuration: lockableDuration,
-					GaugeId:          record.GaugeId,
-				}
-
-				incentivizedPools = append(incentivizedPools, incentivizedPool)
-				incentivizedPoolIDs[poolId] = lockableDuration
-			}
-		}
-	}
-
-	// Only run the following if the above loop determined there were incentivized pools.
-	if len(incentivizedPoolIDs) > 0 {
-		// Retrieve the migration records between balancer pools and concentrated liquidity pools.
-		// This comes from the superfluid keeper, since superfluid is the only pool incentives connected
-		// module that has access to the gamm modules store.
-		migrationRecords, err := q.gammKeeper.GetAllMigrationInfo(sdkCtx)
+		gauge, err := q.incentivesKeeper.GetGaugeByID(sdkCtx, record.GaugeId)
 		if err != nil {
 			return nil, status.Error(codes.Internal, err.Error())
 		}
 
-		// Iterate over all migration records.
-		for _, record := range migrationRecords.BalancerToConcentratedPoolLinks {
-			// If the cl pool is not in the list of incentivized pools, skip it.
-			lockableDuration, incentivized := incentivizedPoolIDs[record.ClPoolId]
-			if !incentivized {
-				continue
+		if gauge.DistributeTo.LockQueryType == lockuptypes.ByGroup {
+			group, err := q.Keeper.incentivesKeeper.GetGroupByGaugeID(sdkCtx, record.GaugeId)
+			if err != nil {
+				return nil, status.Error(codes.Internal, err.Error())
 			}
```

Comment:
We need to make one more check.

Groups can be both perpetual and not. Perpetual are for internal incentives and non-perpetual for external

As a result, please also retrieve a group gauge and confirm that is is indeed perpetual. If not, skip it.

Reply:
I am adding the check but to be clear, I think we have a contract that DistrRecords cannot be non-perpetual to start with

Reply:
See https://github.com/osmosis-labs/osmosis/blob/7dbd9e975185c50e18228f6f94aeb8bcd18aceaf/x/pool-incentives/keeper/distr.go#L114-L125

Reply:
Added this check and revamped the test (previous test made very little sense / was actually testing nothing in many cases) https://github.com/osmosis-labs/osmosis/pull/6601/commits/8775fee04914e4c1dc110ea9c8a10a276814cd21

Reply:
Check added https://github.com/osmosis-labs/osmosis/pull/6601/commits/8775fee04914e4c1dc110ea9c8a10a276814cd21